plaidml/bridge/openvino/ops/deformable_convolution.cpp

```diff
@@ -0,0 +1,280 @@
+// Copyright (C) 2020 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "plaidml_ops.hpp"
+#include "plaidml_util.hpp"
+
+#include "ngraph/opsets/opset.hpp"
+#include "ngraph/opsets/opset4.hpp"
+
+#include "plaidml/op/op.h"
+
+using namespace plaidml;          // NOLINT[build/namespaces]
+using namespace InferenceEngine;  // NOLINT[build/namespaces]
+using namespace edsl;
+
+namespace PlaidMLPlugin {
+
+// Compute pad_before and the size of output
+std::pair<size_t, size_t> compute_padding_before_and_output_size(size_t input_size, size_t off_size, size_t filter_size,
+                                                                 size_t stride, plaidml::op::AutoPadMode autopad_mode,
+                                                                 size_t pad_before, size_t pad_end, size_t dilation) {
+  size_t I_eff = input_size;
+  size_t F_eff = (dilation * (filter_size - 1)) + 1;
+  if (autopad_mode == plaidml::op::AutoPadMode::EXPLICIT) {
+    size_t output_size = ((I_eff + pad_before + pad_end - F_eff + stride) / stride);
+    std::pair<size_t, size_t> result(pad_before, output_size);
+    return result;
+  }
+  if (autopad_mode == plaidml::op::AutoPadMode::VALID) {
+    size_t output_size = ((I_eff - F_eff + stride) / stride);
+    std::pair<size_t, size_t> result(0, output_size);
+    return result;
+  }
+  if (autopad_mode == plaidml::op::AutoPadMode::SAME_LOWER || autopad_mode == plaidml::op::AutoPadMode::SAME_UPPER) {
+    size_t lower_term = (autopad_mode == plaidml::op::AutoPadMode::SAME_LOWER) ? 1 : 0;
+    size_t max = 0;
+    if (((off_size - 1) * stride + F_eff - input_size) > 0) {
+      max = (off_size - 1) * stride + F_eff - input_size;
+    }
+    pad_before = (max + lower_term) / 2;
+    size_t output_size = ((I_eff + stride - 1) / stride);
+    std::pair<size_t, size_t> result(pad_before, output_size);
+    return result;
+  }
+  THROW_IE_EXCEPTION << "Unexpected autopadding mode.";
+}
+
+// Extract values which are needed.
+edsl::Tensor extract_tensor(edsl::Tensor I, int rank) {
```

Comment:
It seems you are selecting a subset of I to O, can you use reshape, slice and other ops instead of using this function?

Reply:
No, i don't think so, for example, in 2d convolution, the input tensor of the function extract_tensor is {N, CI, N, CI, NEWH, NEWW, 2, N, CI, NEWH, NEWW, 2}, and the output tensor is {N, CI, NEWH, NEWW}, the values which we need are intermixed  in input tensor.

Reply:
so if I understand correctly, is the 2d algorithm here implements the following formula?
O( in, ic, ih, iw )= I (in , ic, in, ic, ih, iw, 0, in, ic, ih, iw, 1)

Reply:
yes, it is.

Reply:
I think this function can be rearranged to make clearer, as follows, please check and feel free to make any changes.
```
// rename passed rank to O_rank.
edsl::Tensor extract_tensor(edsl::Tensor I, int O_rank) {
auto I_rank = I.rank();
// check I_rank==O_rank - 2) * (O_rank + 1) + 2;

std::vector<TensorDim> O_dims(O_rank);
std::vector<TensorIndex> O_idxs(O_rank);

std::vector<TensorDim> I_dims(I_rank);
I.bind_dims(I_dims);
for (auto i = 2; i < rank + 2; ++i) {
    O_dims[i - 2] = I_dims[i];
}

std::vector<TensorIndex> I_idxs;
I_idxs.push_back(O_idxs[0]);
I_idxs.push_back(O_idxs[1]);
for(auto i=0; i<rank; i++){
     I_idxs.insert(I_idxs.end(),O_idxs.begin(),O_idxs.end());
     I_idxs.push_back(edsl::TensorIndex(i));
}
```

Reply:
it is clearer than before, thanks.